File: app/schema/menu.py
"""菜单管理 Schema：基于 Permission 模型的菜单树操作。

菜单类型枚举：
- 1=目录（sub-menu group）
- 2=菜单（menu item / 页面）
- 3=按钮/接口（权限码）
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_serializer

from app.core.clock import LOCAL_TZ

# ── 创建 / 更新 ──────────────────────────────────────────────


class MenuCreate(BaseModel):
    """新建菜单（目录、菜单项或按钮）。"""

    name: str = Field(..., min_length=1, max_length=64, description="菜单名称")
    code: str = Field(..., min_length=1, max_length=100, description="权限标识")
    type: int = Field(2, ge=1, le=3, description="类型: 1=目录 2=菜单 3=按钮")
    parent_id: int | None = Field(None, description="上级菜单ID")
    path: str | None = Field(None, max_length=200, description="前端路由路径")
    component: str | None = Field(None, max_length=200, description="前端组件路径")
    icon: str | None = Field(None, max_length=64, description="图标名称")
    sort: int = Field(0, ge=0, description="排序号")
    status: bool = Field(True, description="是否启用")
    redirect: str | None = Field(None, max_length=200, description="默认跳转地址")
    is_hidden: bool = Field(False, description="是否隐藏路由")
    is_cache: bool = Field(False, description="是否缓存路由(KeepAlive)")
    is_affix: bool = Field(False, description="是否聚合路由")
    is_external: bool = Field(False, description="是否外链(外部打开)")
    remark: str | None = Field(None, max_length=255, description="备注")


class MenuUpdate(BaseModel):
    """更新菜单（字段均为可选，仅传需要修改的字段）。"""

    name: str | None = Field(None, min_length=1, max_length=64, description="菜单名称")
    code: str | None = Field(None, min_length=1, max_length=100, description="权限标识")
    type: int | None = Field(None, ge=1, le=3, description="类型")
    parent_id: int | None = Field(None, description="上级菜单ID")
    path: str | None = Field(None, max_length=200, description="前端路由路径")
    component: str | None = Field(None, max_length=200, description="前端组件路径")
    icon: str | None = Field(None, max_length=64, description="图标名称")
    sort: int | None = Field(None, ge=0, description="排序号")
    status: bool | None = Field(None, description="是否启用")
    redirect: str | None = Field(None, max_length=200, description="默认跳转地址")
    is_hidden: bool | None = Field(None, description="是否隐藏路由")
    is_cache: bool | None = Field(None, description="是否缓存路由(KeepAlive)")
    is_affix: bool | None = Field(None, description="是否聚合路由")
    is_external: bool | None = Field(None, description="是否外链(外部打开)")
    remark: str | None = Field(None, max_length=255, description="备注")


# ── 输出 ────────────────────────────────────────────────────


class MenuOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    code: str
    type: int
    parent_id: int | None = None
    path: str | None = None
    component: str | None = None
    icon: str | None = None
    sort: int = 0
    status: bool = True
    redirect: str | None = None
    is_hidden: bool = False
    is_cache: bool = False
    is_affix: bool = False
    is_external: bool = False
    remark: str | None = None
    is_deleted: bool = False
    created_at: datetime | None = None
    updated_at: datetime | None = None

    @field_serializer("created_at")
    def _serialize_created(self, v: datetime | None) -> str | None:
        if v is None:
            return None
        return v.astimezone(LOCAL_TZ).replace(tzinfo=None).isoformat()

    @field_serializer("updated_at")
    def _serialize_updated(self, v: datetime | None) -> str | None:
        if v is None:
            return None
        return v.astimezone(LOCAL_TZ).replace(tzinfo=None).isoformat()


class MenuTreeOut(MenuOut):
    """树形菜单节点，携带下级 children。"""

    children: list["MenuTreeOut"] = []

File: web/src/api/menu.ts
/** 菜单管理 API 封装。
 *
 * 菜单基于 Permission 模型，type=1(目录) / 2(菜单项) / 3(按钮/接口)。
 * 菜单管理页管理全部类型；导航渲染请使用 fetchMenuTree()，它仅返回启用状态的目录和菜单。
 */
import { http } from "@/utils/request";
import { batchDelete, type BatchDeleteResult } from "@/api/batch";

export interface MenuItem {
  id: number;
  name: string;
  code: string;
  type: number; // 1=目录 2=菜单 3=按钮
  parent_id: number | null;
  path: string | null;
  component: string | null;
  icon: string | null;
  sort: number;
  status: boolean;
  redirect: string | null;
  is_hidden: boolean;
  is_cache: boolean;
  is_affix: boolean;
  is_external: boolean;
  remark: string | null;
  is_deleted: boolean;
  created_at: string | null;
  updated_at: string | null;
}

export interface MenuTreeItem extends MenuItem {
  children: MenuTreeItem[];
}

export interface MenuCreate {
  name: string;
  code: string;
  type: number;
  parent_id?: number | null;
  path?: string | null;
  component?: string | null;
  icon?: string | null;
  sort?: number;
  status?: boolean;
  redirect?: string | null;
  is_hidden?: boolean;
  is_cache?: boolean;
  is_affix?: boolean;
  is_external?: boolean;
  remark?: string | null;
}

export interface MenuUpdate {
  name?: string | null;
  code?: string | null;
  type?: number | null;
  parent_id?: number | null;
  path?: string | null;
  component?: string | null;
  icon?: string | null;
  sort?: number | null;
  status?: boolean | null;
  redirect?: string | null;
  is_hidden?: boolean | null;
  is_cache?: boolean | null;
  is_affix?: boolean | null;
  is_external?: boolean | null;
  remark?: string | null;
}

export interface MenuQuery {
  keyword?: string;
  type?: number | null;
  status?: boolean | null;
}

// ── 查询 ──

export function fetchMenus(params: MenuQuery = {}): Promise<MenuTreeItem[]> {
  return http<MenuTreeItem[]>({ url: "/v1/menus", method: "GET", params });
}

export function fetchMenuTree(): Promise<MenuTreeItem[]> {
  return http<MenuTreeItem[]>({ url: "/v1/menus/tree", method: "GET" });
}

export function fetchMenuOptions(): Promise<MenuItem[]> {
  return http<MenuItem[]>({ url: "/v1/menus/options", method: "GET" });
}

export function fetchMenu(id: number): Promise<MenuItem> {
  return http<MenuItem>({ url: `/v1/menus/${id}`, method: "GET" });
}

// ── 写入 ──

export function createMenu(data: MenuCreate): Promise<MenuItem> {
  return http<MenuItem>({ url: "/v1/menus", method: "POST", data });
}

export function updateMenu(id: number, data: MenuUpdate): Promise<MenuItem> {
  return http<MenuItem>({ url: `/v1/menus/${id}`, method: "PUT", data });
}

export function deleteMenu(id: number): Promise<null> {
  return http<null>({ url: `/v1/menus/${id}`, method: "DELETE" });
}

// ── 批量 ──

export function batchDeleteMenus(ids: number[]): Promise<BatchDeleteResult> {
  return batchDelete("/v1/menus/batch-delete", ids);
}
